### PR TITLE
Unexclude _utils/, exceptions.py, and _exception_helpers

### PR DESCRIPTION
**Pull request by CoderKevinZhang**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py [hunk 1]
```diff
@@ -113,8 +113,8 @@ def __init__(
         message: str,
         no_personal_data_message: str,
         *args,
-        target: ErrorTarget = ErrorTarget.UNKNOWN,
-        error_category: ErrorCategory = ErrorCategory.UNKNOWN,
+        target: Union[str, ErrorTarget, ErrorCategory] = ErrorTarget.UNKNOWN,
+        error_category: Union[str, ErrorTarget, ErrorCategory] = ErrorCategory.UNKNOWN,
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:117]: Why is this ErrorCategory and ErrorTarget? It should only be the corresponding one for each

```python
target: Union[str, ErrorTarget] = ErrorTarget.UNKNOWN,
error_category: Union[str, ErrorCategory] = ErrorCategory.UNKNOWN,

```

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py [hunk 2]
```diff
@@ -214,17 +212,15 @@ def __init__(
         message: str,
         no_personal_data_message: str,
         *args,
-        target: ErrorTarget = ErrorTarget.UNKNOWN,
-        error_category: ErrorCategory = ErrorCategory.UNKNOWN,
+        target: Union[str, ErrorTarget] = ErrorTarget.UNKNOWN,
+        error_category: Union[str, ErrorTarget] = ErrorCategory.UNKNOWN,
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:216]: Again, these categories should match.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py [hunk 3]
```diff
@@ -214,17 +212,15 @@ def __init__(
...
         **kwargs,
     ):
         super(ComponentException, self).__init__(
             message=message,
+            no_personal_data_message=no_personal_data_message,
             target=target,
             error_category=error_category,
-            no_personal_data_message=no_personal_data_message,
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:225]: why remove the args and kwargs?

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py [hunk 4]
```diff
@@ -247,17 +243,15 @@ def __init__(
         message: str,
         no_personal_data_message: str,
         *args,
-        target: ErrorTarget = ErrorTarget.UNKNOWN,
-        error_category: ErrorCategory = ErrorCategory.UNKNOWN,
+        target: Union[str, ErrorTarget] = ErrorTarget.UNKNOWN,
+        error_category: Union[str, ErrorTarget] = ErrorCategory.UNKNOWN,
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:247]: categories should match

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py [hunk 5]
```diff
@@ -247,17 +243,15 @@ def __init__(
...
         **kwargs,
     ):
         super(JobException, self).__init__(
             message=message,
+            no_personal_data_message=no_personal_data_message,
             target=target,
             error_category=error_category,
-            no_personal_data_message=no_personal_data_message,
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:258]: why args and kwargs? same comment for all classes in this file.

File: sdk/ml/azure-ai-ml/pyproject.toml [hunk 6]
```diff
@@ -26,14 +26,12 @@ exclude = [
   "azure/ai/ml/_vendor/",
   "azure/ai/ml/_schema/",
   "azure/ai/ml/_arm_deployments/",
+  "azure/ai/ml/_utils/_virtual_cluster_utils/__init__.py",
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/pyproject.toml:29]: Why do we have to exclude this file?

**Review by diondrapeck — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py [hunk 7]
```diff
@@ -521,9 +525,9 @@ def upload_file(
     ):  # Only for Gen2StorageClient, Blob Storage doesn't have true directories
         if in_directory:
             storage_client.temp_sub_directory_client = None
-            file_name_tail = dest.split(os.path.sep)[-1]
+            file_name_tail = str(dest).rsplit(os.path.sep, maxsplit=1)[-1]
```
[the remark this answers is not in the document]
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py:528]: this is auto-recommended by mypy check in order to remove the mypy error

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py [hunk 8]
```diff
@@ -769,7 +773,7 @@ def from_iso_duration_format_min_sec(duration: Optional[str]) -> str:
     :return: The converted duration
     :rtype: str
     """
-    return duration.split(".")[0].replace("PT", "").replace("M", "m ") + "s"
+    return str(duration).split(".", maxsplit=1)[0].replace("PT", "").replace("M", "m ") + "s"
```
[the remark this answers is not in the document]
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py:776]: this is auto-recommended by mypy check in order to remove the mypy error

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 1 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:117]: image

ErrorCategory type is passed into target above

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 2 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:216]: mypy check is considered `ErrorTarget.UNKNOWN` as `str` type, not `ErrorTarget `type

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 3 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:225]: image

mypy would complain about this error

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 4 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:247]: same above

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 5 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:258]: same above

File: sdk/ml/azure-ai-ml/pyproject.toml (hunk 6 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/pyproject.toml:29]: https://github.com/Azure/azure-sdk-for-python/pull/33860/files#diff-af1044a8cc549e69969dc610c35f3658a5ff6aa94a12162855fbfa2574d312e5

This is a closed PR that I follow your change and you exclude it in this PR

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py:528]: Based on reading the documentation for rsplit and the maxsplit parameter, this makes me nervous. Let's revert to the original code and ignore mypy here for that specific error.

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py:776]: same as above - let's revert and ignore.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 1 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:117]: Okay, that's a bug. Let's change the type in DeploymentException for target to Union[str, ErrorTarget] and that should resolve the issues for MlException.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 2 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:216]: That's fine. I should've been clearer. I meant that `error_category: Union[str, ErrorTarget]` should be `error_category: Union[str, ErrorCategory]`. Fixing the bug in DeploymentException's type annotations should make this clear and then make the mypy errors go away.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 3 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:225]: Let's ignore those mypy errors for now.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 4 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:247]: Same comment as above, let's fix these all over. Once the bug above is resolved, we should be good to make `error_category: Union[str, ErrorCategory]` everywhere.

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 5 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:258]: same as above, let's ignore mypy [misc] here

File: sdk/ml/azure-ai-ml/pyproject.toml (hunk 6 above)
Comment (diondrapeck) [reply sdk/ml/azure-ai-ml/pyproject.toml:29]: My apologies for the confusion. I believe I just just testing something with that commit. Let's uncomment it and let me know if any mypy errors surface.

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py:528]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/utils.py:776]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 1 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:117]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 3 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:225]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 5 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:258]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 2 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:216]: fixed

File: sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py (hunk 4 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/exceptions.py:247]: fixed

File: sdk/ml/azure-ai-ml/pyproject.toml (hunk 6 above)
Comment (CoderKevinZhang) [reply sdk/ml/azure-ai-ml/pyproject.toml:29]: new mypy errors occur when excluding this

image

**Pull request closed without merging**